### PR TITLE
hw 4 added

### PR DESCRIPTION
**Pull request by nirtake**

### CODE REVIEW

**Review by fortvivlan — COMMENTED** [review]
```
data_transforms = {
    'train': transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize(mean, std)
    ]),
    'test': transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize(mean2, std2)
    ]),
}
```
К тестовым данным надо применять ту же нормализацию, что и к обучающим - по аналогии с mean/std в StandardScaler или BatchNorm. Плюс тестовые данные могут не быть доступны все сразу (т.н. online постановка задачи).
К тому же у вас проиходит утечка данных в валидацию: надо сначала разбить на train/val, и только потом на трейне считать поканальные mean/std, а не наоборот.

---

`val_accuracy.extend((torch.argmax(y_pred, dim=-1) == y_val).cpu().numpy().tolist())`
Такие вещи довольно плохо читаются, лучше разбить на несколько более понятных пунктов, вроде `y_preds = torch.argmax(y_logits, dim=-1); ...`

---

```
        self.conv1 = nn.Conv2d(3, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.pool = nn.MaxPool2d(2)
        self.calculate_flatten_size()
        self.fc1 = nn.Linear(self.flatten_size, 128)
        self.fc2 = nn.Linear(128, num_classes)
```
->
```
nn.Sequential(
  nn.Conv2d(3, 10, kernel_size=5),
  nn.Conv2d(10, 20, kernel_size=5)
  nn.MaxPool2d(2),
  nn.Flatten(),
  nn.Linear(N, 128),
  nn.Linear(128, num_classes)
)
```
а N посчитать ручками. По-моему не нужно усложнять, если можно сделать просто

---

```
Epoch: 10	loss: 1.1896130868366785	accuracy: 0.705607476635514
Epoch: 20	loss: 1.4191671184131078	accuracy: 0.719626168224299
```
Видно же, что оно переобучается... Нужно было логировать лосс каждую эпоху и учить максимум 20 эпох, найти наилучшую эпоху и уже потом менять параметры и что бы то ни было. А еще лучше запоминать веса наилучшей эпохи в процессе обучения (i.e. best_model = model if current_val_loss < best_val_loss else best_model), тогда не нужно будет каждый раз руками подбирать максимальное количество эпох.
